### ISSUE TITLE
LIFO Channel Select Order in Sequential Functions

### ISSUE THREAD

**Issue by DrGermanius** [vlang/v#18553]
### Describe the bug

The bug happens when executing multiply functions consecutively, each containing a select statement that listens on the same channel. Despite the order of execution, the value written to the channel consistently gets processed by the select statement launched later, rather than following a first-in-first-out (FIFO) pattern as expected.

### Expected Behavior

I expected the behavior to be similar to Go (FIFO). The first select will be processed.

https://go.dev/play/p/y8R3BdVjv6f

### Current Behavior

The latest select processed.

### Reproduction Steps

```
module main

import time

fn f(ch chan int) {
	select {
		item := <-ch {
			println(item)
		}
	}
	println('first func')
}

fn s(ch chan int) {
	select {
		item := <-ch {
			println(item)
		}
	}
	println('second func')
}
fn main() {
	ch := chan int{}
	spawn f(ch)
	time.sleep(1 * time.second)
	spawn s(ch)
	time.sleep(1 * time.second)
	ch <- 999
	time.sleep(1 * time.second)
}

```

### V version

V 0.3.4 af4e113

### Environment details (OS name and version, etc.)

```
OS: macos, macOS, 13.2.1, 22D68
Processor: 12 cpus, 64bit, little endian, Apple M2 Max
CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20
```

**Comment by DrGermanius** [comment]
@JalonSolov I expected that the first select will be processed firstly. So in provided code I expect to see "first". 
As I understood LIFO ordering for concurrent selects is not a special feature and this was done unintentionally.
So now my main question is should this behavior be fixed or LIFO is ok?

**Comment by JalonSolov** [comment]
You said you wanted FIFO order, which means the first one you spawned should have seen the first value on the channel... which is what I saw.  The `999` was printed by `f()`, while `second func` was printed by the `s()`.

As for which it should be?  You're the first that has asked for a specific order.  If we allow choosing a specific order, I would say it should be a user setting, with perhaps 3 options:

RANDOM
FIFO
LIFO

(with RANDOM as the default - what we have now)

However, it's easier for you to force the order you want, than for this to be implemented in the compiler.

**Comment by DrGermanius** [comment]
@JalonSolov 
try to run following:

```
module main

import time

fn f(ch chan int, text string) {
	select {
		item := <-ch {
			//println("")
		}
	}
	println(text)
}

fn main() {
	for {
		time.sleep(3 * time.second)
		ch := chan int{}
		spawn f(ch, "1")
		time.sleep(2 * time.second)
		spawn f(ch, "2")
		time.sleep(2 * time.second)
		spawn f(ch, "3")
		time.sleep(2 * time.second)
		ch <- 999
		ch <- 999
		ch <- 999
		time.sleep(3 * time.second)
		println("")
		println("")
	}
}

```

My Output:
```
3
2
1


3
2
1


3
2
1


3
2
1


3
2
1


3
2
1


3
2
1
```

The long sleeps are needed for corectesses (all gorutines are starting and so on). 
So looks like it not randomly . btw please read #7415 where Use saying that it's `Actually it's even a kind of "last come, first served"`

I'm seeing that sometimes it can produce "3 1 2". But in the most cases it's just LIFO. Also tried with 5 functions and got the same behavior. So maybe it should be _**true**_ random if you are saying that it's a default option?

**Comment by DrGermanius** [comment]
Yeah, it's really random when we talk about spawning system threads, but I asked about something else. If we have 3 threads with selects that listen to the same channel, and the order of spawning these threads will be `1 - 2 - 3`, the order to receive the value from the channel will always be `3 - 2 - 1 `(LIFO, now it's implemented using some kind of linked list). So I'm asking - was that done for a specific reason or unintentionally?

**Comment by JalonSolov** [comment]
I'm saying it is happening unintentionally because it is happening in random order, even if it _appears_ to be happening in LIFO order... most of the time (but not always).

**Comment by JalonSolov** [comment]
Kinda defeats the purpose of "asynchronous", but whatevs.

**Comment by DrGermanius** [comment]
I completely agree with @medvednikov. I firmly believe it should operate in a manner similar to GO. Here is the script that demonstrates the FIFO order. https://go.dev/play/p/n4OaP4dpzKf

**Comment by Casper64** [comment]
@DrGermanius I noticed that in applications where multiple threads are selecting from the same channel, the CPU usage is high for the first few threads and very low for the last threads. As if they are not receiving the equal amount of work.

Could this also explained by the current workings of channels?

**Comment by JalonSolov** [comment]
Another problem with forcing FIFO... if one thread gets "stuck", it freezes the whole chain.  If the runtime is "smart" enough, it can move on to the next thread, to avoid long delays, but that would defeat FIFO.

There's a _big_ difference between FIFO on channel data (which is handled correctly right now) and FIFO on which thread runs.

**Comment by DrGermanius** [comment]
@JalonSolov we are talking about FIFO for getting values from channels.

**Comment by DrGermanius** [comment]
@Casper64 I'm not sure. In both languages, there are linked lists of writers and listeners. In the Go implementation, the address of the first channel is picked up from that list, and the value is sent to that address. In V, it seems to work similarly, but the linked list is reversed (although the reason for this is unknown). Consequently, if nothing is written to the channel, no action is taken with the threads.

**Comment by JalonSolov** [comment]
Then your example is flawed.  You're only printing a value for which thread is run, not the data pulled from the channel.  Using this modification, it appears that the data on the channel _is_ handled in FIFO order already:
```v
module main

import time

fn f(ch chan int, text string) {
        select {
                item := <-ch {
                        println('${text}: ${item}')
                }
        }
}

fn main() {
        for {
                time.sleep(3 * time.second)
                ch := chan int{}
                spawn f(ch, "1")
                time.sleep(2 * time.second)
                spawn f(ch, "2")
                time.sleep(2 * time.second)
                spawn f(ch, "3")
                time.sleep(2 * time.second)
                ch <- 1
                ch <- 2
                ch <- 3
                time.sleep(3 * time.second)
                println("")
                println("")
        }
}
```
output is:
```
3: 1
2: 2
1: 3


3: 1
2: 2
1: 3


...
```

**Comment by DrGermanius** [comment]
BTW I had an error in my V script (missing sleeps)

So it works **always** in LIFO way.

```
module main

import time

fn f(ch chan int, text string) {
	println('runned: ' + text)
	select {
		item := <-ch {
			//println("")
		}
	}
	println(text)
}

fn main() {
	for {
		time.sleep(3 * time.second)
		ch := chan int{}
		spawn f(ch, "1")
		time.sleep(2 * time.second)
		spawn f(ch, "2")
		time.sleep(2 * time.second)
		spawn f(ch, "3")
		time.sleep(1 * time.second)
		ch <- 999
		time.sleep(1 * time.second)
		ch <- 999
		time.sleep(1 * time.second)
		ch <- 999
		time.sleep(3 * time.second)
		println("")
		println("")
	}
}

```

**Comment by DrGermanius** [comment]
@JalonSolov in your example the third thread always gets the first value. This is LIFO.

**Comment by JalonSolov** [comment]
You said this wasn't about the thread order, just the data order.  My example shows the data is handled in FIFO order.

**Comment by DrGermanius** [comment]
Here is FIFO: https://go.dev/play/p/mhz-koVoZd3

**Comment by DrGermanius** [comment]
@JalonSolov please read about how channels works in my answer to @Casper64.

**Comment by DrGermanius** [comment]
> You said this wasn't about the thread order, just the data order. My example shows the data is handled in FIFO order.
It's about the order of channels to receive data.

**Comment by JalonSolov** [comment]
You only have one channel, and the data sent to that channel is being handled in FIFO order now.

Threads and channels are 2 separate things.

**Comment by DrGermanius** [comment]
I don't think it's productive to argue about words. There are two approaches: GO and V. I think V channels must work in the same way as Go channels. 
So by channels I mean the "address of writer/receiver".

**Comment by JalonSolov** [comment]
So I did go to the Go playground page, and I executed the sample code there.  This was the output:
```
999
first func
```
... which is LIFO order for the spawned routines, not FIFO.  Otherwise, it would have shown
```
999
second func
```
Since `f()` was spawned first, and `s()` was spawned 2nd.

Note that when I ran your example the first time, I saw the code ran in FIFO order, giving the expected output of
```
999
second func
```
However, in the later example, we saw the threads executing in LIFO order, but the data was, indeed, handled in FIFO order.

So please clarify... what _exactly_ are you requesting?  That the threads always execute in FIFO order, as well as handling the data in FIFO order?

The spawn command creates system threads, and the execution order of those threads cannot be determined ahead of time, unless extra code is added to _force_ them to run in a particular order.
```
The exact timing of when tasks in a concurrent system are executed depends on the scheduling, and tasks need not always be executed concurrently. For example, given two tasks, T1 and T2:

        T1 may be executed and finished before T2 or vice versa (serial and sequential)

        T1 and T2 may be executed alternately (serial and concurrent)

        T1 and T2 may be executed simultaneously at the same instant of time (parallel and concurrent)
```
Running coroutines (same thing as goroutines, basically, and invoked with the `go` keyword in both Go and V), however, could be made to run in FIFO order, since they require a run-time scheduler.  I don't know if the scheduler V has adopted (photon) enforces FIFO order or not.

**Comment by DrGermanius** [comment]
```
module main

import time

fn f_no_select(ch chan string, text string){
	v := <-ch
	println('$text got $v')
}


fn main() {
	ch := chan string{}
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'first thread')
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'second thread')
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'third thread')
	time.sleep(1 * time.second)
	ch <- 'first value'
	time.sleep(1 * time.second)
	ch <- 'second value'
	time.sleep(1 * time.second)
	ch <- 'third value'
	time.sleep(1 * time.second)
}
```

output: 

```
first thread got first value
second thread got second value
third thread got third value
```

Each thread subscribes to the channel. The value is passed firstly to the first (the most recent) subscriber, then to the second subscriber, and so on. I call this the 'FIFO order for processing the channel listeners'.

However, if I simply add a select statement, the situation will be reversed.

```
module main

import time

fn f_no_select(ch chan string, text string){
	select {
		v := <-ch{
			println('$text got $v')
		}
	}
}


fn main() {
	ch := chan string{}
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'first thread')
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'second thread')
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'third thread')
	time.sleep(1 * time.second)
	ch <- 'first value'
	time.sleep(1 * time.second)
	ch <- 'second value'
	time.sleep(1 * time.second)
	ch <- 'third value'
	time.sleep(1 * time.second)
}

```

output: 
```
third thread got first value
second thread got second value
first thread got third value
```

I'm requesting consistency in the execution of the aforementioned tasks. That means that the first thread which subscribed to the channel inside the select statement will receive the value first. My point is that they should operate in the same manner as they do in GO.

It is not about the order of spawning threads; as you mentioned, it is random. However, I use time.sleep(1 * time.second) to avoid that randomness.

**Comment by JalonSolov** [comment]
So the only actual issue here is that `select` appears to be flipping the order from FIFO to LIFO.

Question:  Why use select?  I agree it probably shouldn't be flipping the order, but I'm wondering why it's needed, since the code appears to work fine without it.

**Comment by DrGermanius** [comment]
It's just a small example. I was working on a queue where users can subscribe to a specific queue with a timeout. Here is the description:

"If there are no ready messages in the queue, the recipient will wait either until a message arrives or until the timeout expires. If the timeout is reached and no message appears, the response will be a status code 404 (Not Found).

Note: Recipients receive messages in the same order as their requests were made. If multiple recipients are waiting for a message using the timeout, the first recipient to make the request will receive the first message."

My first implementation was as follows. A select statement is needed for implementing the timeout logic

```
fn (mut q Queue) wait(timeout_sec int, q_name string) !string {
 	ch := q.m[q_name] or { return NotFoundError{} }
 	select {
 		item := <-ch {
 			return item
 		}
 		timeout_sec * time.second {
			return NotFoundError{}
		}
	}
	return NotFoundError{}
}

```

Then I realized that users were receiving messages in LIFO order, meaning that the last subscribed user would get the first message. I've implemented a workaround by using a slice of channels but anyway, I believe, that logic of selects should be fixed (at least because in GO that works like common channels).

**Comment by DrGermanius** [comment]
While fixing this behavior, I discovered another issue that appears to be more significant. If I execute two functions, one with a select statement and the other without it (in any order), the order of execution will be as follows: 1) functions without select, 2) function with select. This occurs due to the presence of two separate queues (one for select statements and another for non-select statements) within the channel type, unlike in the GO language where only one queue is used.

```
module main

import time

fn f_with_select(ch chan string, text string) {
	select {
		v := <-ch {
			println('${text} got ${v}')
		}
	}
}

fn f_no_select(ch chan string, text string) {
	v := <-ch
	{
		println('${text} got ${v}')
	}
}

fn main() {
	ch := chan string{}
	time.sleep(1 * time.second)
	spawn f_with_select(ch, 'select')
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'no')
	time.sleep(1 * time.second)
	spawn f_with_select(ch, 'select')
	time.sleep(1 * time.second)
	spawn f_no_select(ch, 'no')
	time.sleep(1 * time.second)
	ch <- 'first value'
	time.sleep(1 * time.second)
	ch <- 'second value'
	time.sleep(1 * time.second)
	ch <- 'third value'
	time.sleep(1 * time.second)
	ch <- 'fourth value'
	time.sleep(1 * time.second)
}

```

output: 
```
no got first value
no got second value
select got third value
select got fourth value
```

**Comment by peterbourgon** [comment]
> The bug happens when executing multiply functions consecutively, each containing a select statement that listens on the same channel. Despite the order of execution, the value written to the channel consistently gets processed by the select statement launched later, rather than following a first-in-first-out (FIFO) pattern as expected.

There seems to be a fundamental misunderstanding here, sleeping between spawns doesn't guarantee any specific execution order, the scheduler is free to run the last spawned task before the first one, or any other order it chooses. There is no well-defined "order of execution" of spawned tasks.

FIFO is a property of a channel as a data structure, not of spawned threads/goroutines which write to and/or read from a channel. If you have one writer and multiple concurrent readers on a single channel then you won't necessarily observe FIFO behavior.

```go
fn main() {
	ch := chan int{}
	spawn f(ch)
	time.sleep(1 * time.second)
	spawn s(ch)
	time.sleep(1 * time.second)
	ch <- 999
	time.sleep(1 * time.second)
}
```

Whether 999 is received by the f task or the s task is arbitrary, either outcome is valid, and neither outcome is guaranteed. In fact it's entirely possible that the program exits before 999 is received by either of those tasks. 

> Each thread subscribes to the channel. The value is passed firstly to the first (the most recent) subscriber, then to the second subscriber, and so on. I call this the 'FIFO order for processing the channel listeners'.

This isn't correct, the value is passed to a random "subscriber", the order in which they "subscribed" doesn't matter.

**Comment by DrGermanius** [comment]
No, @peterbourgon, it does not passing to a random subscriber. You can check the channels.v code by yourself or you can rely on https://github.com/vlang/v/pull/7415, where Uwe(creator of this logic) said about LIFO ordering. 


Talking about order of threads execution - I agree with you about scheduler, but using time.sleep we can determine the order **_in most of_** the cases. 

Also you can check this [message](https://github.com/vlang/v/issues/18553#issuecomment-1614180375), I believe it contains the better example of problem.
